support: add interfaces to use custom seeds

tylershunt opened the pull request:
Call `GetGenerator(seed)` to get a randomness generator with your seed
of choice. This can then be passed back to the other utils as the last
parameter. If omitted, utils behave as before.

Signed-off-by: Tyler Hunt <thunt@katanagraph.com>

ddn0 commented on libsupport/CMakeLists.txt:
```diff
@@ -21,6 +22,17 @@ if (fmt_VERSION VERSION_LESS 4)
+  # Autogenerated conan module doesn't provide header-only target. Extract one
+  # manually.
+  get_target_property(include_dirs Boost::Boost INTERFACE_INCLUDE_DIRECTORIES)
+  target_include_directories(galois_support PUBLIC ${include_dirs})
```
I think we'll need
```
target_link_libraries(galois_support PUBLIC Boost::Boost)
```
here

In conan there is only one target: `Boost::Boost` (the library and its header files). In `FindBoost` (the standard cmake config for a system-level boost library), there is `Boost::boost` (the header only library) and the various component libraries (`Boost::filesystem`, etc.).

tylershunt replied:
Yeah fixing that now; the build failed so I'm sure I was working toward that conclusion (eventually ;) )

ddn0 commented on libtsuba/src/s3.cpp:
```diff
@@ -16,6 +16,8 @@
+#include "galois/FileSystem.h"
+#include "galois/Result.h"
 #include "tsuba_internal.h"
 #include "s3.h"
```
while you are here, do you mind moving `#include "s3.h"` to the top (to ensure that "s3.h" includes or forward declares everything it needs)?

tylershunt replied:
you got it.

ddn0 commented on libsupport/src/FileSystem.cpp:
```diff
@@ -0,0 +1,53 @@
+
+galois::Result<std::pair<std::string, int>>
+galois::OpenUniqueFile(std::string_view prefix, std::string_view suffix) {
+  std::vector<char> buf(std::move(TemplateString(prefix, suffix)));
```
I believe `std::move` is not necessary here since C++17 made return-value elision mandatory: https://en.cppreference.com/w/cpp/language/copy_elision

https://godbolt.org/z/tZZmmX

vs

https://godbolt.org/z/xfbext

tylershunt replied:
yeah clang clued me in with their more verbose warnings when I uploaded.

Commit:
support: add interfaces to use custom seeds

Call `katana::RandGenerator(seed)` to get a randomness generator with
your seed of choice. This can then be passed back to the other utils as
the last parameter. If omitted, utils behave as before.

Signed-off-by: Tyler Hunt <thunt@katanagraph.com>

witchel commented on libsupport/test/random.cpp:
```diff
@@ -18,6 +19,34 @@ main() {
+    KATANA_LOG_VASSERT(
+        val == first_val,
+        "seeded rngs should output the same values, "
+        "got \"{}\" and \"{}\"",
```
Oh, so all threads got the same string because they have thread-local generators that are identical?

tylershunt replied:
Almost, in this case I created my own thread local generators which should all produce the same strings by seeding 128 of them with the same value.

witchel reviewed (approved):
Randomness' work is never done.

This is all to allow user to pass in constant generators so they can get deterministic random numbers, right?  Too bad it is so complicated.  But a one-and-done effort I presume.

tylershunt commented:
> This is all to allow user to pass in constant generators so they can get deterministic random numbers, right? Too bad it is so complicated. But a one-and-done effort I presume.

I wanted to support calling these functions on generators that are deterministic for testing, while at the same time leaving randomness intact when it's needed.

For example, in the import work I added a testing feature to add some random length string properties quickly in memory; I wanted an option to make those lengths deterministic without forcing `tsuba` to use a deterministic generator which would cobber property files written by multiple hosts.

ddn0 commented on libsupport/include/katana/Random.h:
```diff
@@ -8,15 +8,27 @@
+
+/// Range 0..len-1, inclusive
+KATANA_EXPORT int64_t
+RandomUniformInt(int64_t len, RandGenerator* gen = nullptr);
```
I don't know how I feel about this :D 

- These generators go to seed per-thread state so it is easy to not use this interface correctly
- Re: `GetGenerator(uint64_t)` part of the reason for taking over generator creation is that a `mt19937` has 19937 bits of state so seeding with an `int64_t` doesn't have a ton of entropy

Could you give me a sense of why you need this functionality?

We already have `RandGenerator& GetGenerator`. Could clients just reset that to whatever they want with some documentation about how this is per-thread?

If the ask is more specific (deterministic mode, testing only), there also might be a cleaner interface.

tylershunt replied:
The ask is a deterministic mode for testing. I want optional determinism for some data that I'm generating on the fly while leaving the random behavior intact to keep things like storage collisions from happening.

ddn0 replied:
I see.

I suppose there is some consistency benefit with using `RandomUniformInt` over `std::uniform_int_distribution` and `RandomAlphanumericString` doesn't have a standard analogue.

Do we really need `GetGenerator(uint64_t)` though?

It's not really the right way to initialize a `std::mt19937`. It is fine to me if users want to opt into taking some shortcuts `GetGenerator() = RandGenerator(seed)`, but I don't think libsupport should imply that this is a common or equivalent way to initialize a generator compared to the default `GetGenerator()`.

tylershunt replied:
Word. Yeah; I'll ditch `GetGenerator(uint64_t)` and users that care can bring their own.

tylershunt commented on libsupport/include/katana/Random.h:
```diff
@@ -8,15 +8,26 @@
+KATANA_EXPORT int64_t
+RandomUniformInt(int64_t len, RandGenerator* gen = nullptr);
+
+/// Range min+1..max-1, inclusive
```
fwiw; this is confusing to me since it opens up corner cases like `min = max - 1` and it seems inconsistent since the float versions are inclusive; could we make them all inclusive? ... but then they behave just like `std::uniform_int_distribution` so maybe the real answer is to ditch these?

witchel replied:
I don't think I knew about `std::uniform_int_distribution` when I wrote this.  That looks fine.

tylershunt commented:
I eliminated everything but the random string generator. See if you still like it (or at least that you don't hate it more).

witchel commented on libtsuba/src/FaultTest.cpp:
```diff
@@ -86,7 +88,9 @@ tsuba::internal::PtP(
+    std::uniform_real_distribution<float> dist(
+        0, std::nextafter(1.0F, std::numeric_limits<float>::max()));
+    if (dist(katana::GetGenerator()) < threshold) {
       fmt::print("  PtP count {:d}\n", ptp_count_);
```
Whoa,` std::nextafter`.  Cool!

witchel commented:
> I eliminated everything but the random string generator. See if you still like it (or at least that you don't hate it more).

Looks good to me.  I assume creating a `std::uniform_int_distribution` doesn't have any hidden costs, like creating the generator.

ddn0 commented on libsupport/include/katana/Random.h:
```diff
@@ -8,15 +8,15 @@
+KATANA_EXPORT std::string RandomAlphanumericString(
+    uint64_t len, RandGenerator* gen = nullptr);
+
+KATANA_EXPORT RandGenerator& GetGenerator();
```
Can you add a comment here that mentions that this returns a thread-local generator?

tylershunt replied:
Will do.

tylershunt commented:
> > I eliminated everything but the random string generator. See if you still like it (or at least that you don't hate it more).
> 
> Looks good to me. I assume creating a `std::uniform_int_distribution` doesn't have any hidden costs, like creating the generator.

I don't think so; but we were constructing a new one for every call to these functions anyway; so if it does this should cost the same as before.
